File: Python3/783. Minimum Distance Between BST Nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def minDiffInBST(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        values = []
        def recursion(node):
            if node:
                recursion(node.left)
                values.append(node.val)
                recursion(node.right)
        def difference(values):
            return abs(values[0] - values[1])
        recursion(root)
        return min(map(difference, zip(values[:-1], values[1:])))

File: Golang/783. Minimum Distance Between BST Nodes.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDiffInBST(root *TreeNode) int {
    values := make([]int, 0, 100)
    recursion(root, &values)
    ans := abs(values[0], values[1])
    for i := 0; i < len(values) - 1; i++ {
        difference := abs(values[i], values[i+1])
        if ans > difference {
            ans = difference
        }
    }
    return ans
}

func recursion(node *TreeNode, values *[]int) {
    if node != nil {
        recursion(node.Left, values)
        *values = append(*values, node.Val)
        recursion(node.Right, values)
    }
}

func abs(a, b int) int {
    if a > b {
        return a - b
    } else {
        return b - a
    }
}
